irc: implement CASEMAPPING parameter for ISUPPORT

Exirel opened the pull request:
### Description

This PR contains various changes with the end goal to use the server's casemapping instead of the hardcoded one. Sadly, this includes breaking changes.

#### Non-breaking changes

* The bot has a new method `make_identifier`: it uses the `CASEMAPPING` parameter from isupport to select the appropriate casemapping function (from `sopel.tools.identifiers`)
* The `Identifier` class has several changes:
  * `Identifier` now lives its best life in `sopel.tools.identifiers`, mostly to prevent cyclic import errors. This comes with some more documentation.
  * `Identifier` doesn't override `str.__new__` anymore, I figured out a way to use `__init__` instead, and it works fine.
  * `Identifier` has a new argument: `casemapping`. It's a callable that **must** take one parameter (a string) and return a string.
  * The default value of `casemapping` is `rfc1459_lower`, making this argument optional (at the moment).
  * `self.casemapping` is now used in place of `Identifier._lower` to get a lowercase version of the identifier string
* Classes that create `Identifier` have a new argument: It's a callable that **must** take one parameter (a string) and return an instance of `Identifier` with properly configured casemapping.
  * `User` and `Channel` (both from `sopel.tools.target`)
  * `PreTrigger| (from `sopel.trigger`)
  * `SopelDB` (from `sopel.db`)
  * `SopelIdentifierMemory` (from `sopel.tools`)
* The new argument is used to instantiate `Identifier` when necessary
* `coretasks` and other modules are now using `bot.make_identifier` when they need an `Identifier`
* All method from `SopelDB` now accept a `str` (some used to accept `Identifier` only) since it can always convert them into an `Identifier` when necessary
* Various type hint here and there because why not

#### Breaking changes

* `Identifier._lower` doesn't lower all character as it used to do, now it uses the `rfc1459_lower` function which lowercase only [A-Z] characters. In theory, this is not a breaking change because IRC identifiers shouldn't accept non-ASCII characters anyway. However, in the case of IRC server supporting Unicode identifiers (for nicknames and/or channels), this is a breaking change. A solution to that would be to implement rfc7613 (PRECIS) as suggested by https://modern.ircdocs.horse/index.html#casemapping-parameter
* `core.nick` is now a `str` and not an `Identifier` anymore, meaning that any plugin using that config option will need to transform it into an `Identifier`; this is considered a breaking change, even tho it shouldn't be a problem considering that most Identifier-related operations are through objects that already convert `str` into `Identifier` when necessary

#### Going further

The next steps would be:

* to merge #2232
* to implement a PRECIS profile for casemapping—I think it would even work on current ascii-only identifier, making Sopel UTF-8 ready before servers can be
* to move memory classes into their own submodule
* and probably to move other things outside of `sopel/tools/__init__.py`
* to implement dynamic chantype for identifiers

### Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

dgw commented on sopel/db.py:
```diff
@@ -393,7 +401,7 @@ def set_nick_value(self, nick, key, value):
             self.ssession.remove()
 
-    def delete_nick_value(self, nick, key):
+    def delete_nick_value(self, nick: str, key: str):
```
Don't forget to remove the redundant types in the `:param:` declarations below.

dgw commented on sopel/db.py:
```diff
@@ -312,7 +322,7 @@ def get_nick_id(self, nick, create=True):
             self.ssession.remove()
 
-    def alias_nick(self, nick, alias):
+    def alias_nick(self, nick: str, alias: str):
```
Don't forget redundant type declarations in `:param:`s below.

dgw commented on sopel/db.py:
```diff
@@ -481,7 +493,7 @@ def unalias_nick(self, alias):
             To *add* an alias for a nick, use :meth:`alias_nick`.
```
…which got type hints added to its definition, but `unalias_nick` didn't.

Neither did `forget_nick_group` and several other related functions below. I don't understand why some touched functions have hinting added and some don't.

Exirel replied:
Yes that's not very thorough: I tend to type hint what I need to understand or modify, and not everything. Given how much I've changed in SopelDB, I'll go through every methods to fix that.

dgw commented on sopel/db.py:
```diff
@@ -896,7 +906,12 @@ def forget_plugin(self, plugin):
     # NICK AND CHANNEL FUNCTIONS
 
-    def get_nick_or_channel_value(self, name, key, default=None):
+    def get_nick_or_channel_value(
+        self,
+        name: Identifier,
+        key: str,
+        default=None
+    ):
```
Redundant (and at least one mismatched) types in `:param:` declarations below.

dgw commented on sopel/db.py:
```diff
@@ -921,7 +936,13 @@ def get_nick_or_channel_value(self, name, key, default=None):
         """
-        name = Identifier(name)
+        if not isinstance(name, Identifier):
+            # TODO: log warning/debug?
+            # The goal is to call this method with an existing Identifier
+            # and not a simple string. This is a convenient/shortcut method
+            # and should probably not be used with a simple string anyway.
```
Is there a reason it shouldn't be callable with a simple `str`? The standalone methods allow it. For API consistency I would _really_ rather not have this be the only place where passing a `str` instead of `Identifier` throws an error, or even a warning.

Exirel replied:
Yeah you are right, now that SopelDB has an identifier maker, that's doable.

dgw commented on sopel/tools/target.py:
```diff
@@ -57,29 +67,35 @@ class Channel:
-        :mod:`sopel.plugin`.
+        This maps nickname :class:`~sopel.tools.identifiers.Identifier`\\s to
+        bitwise integer values. This can be compared to appropriate constants
+        from :mod:`sopel.plugin`.
```
As long as you're here?

```suggestion
        from :mod:`sopel.privileges`.
```

(Which now that I think about it, why isn't that `sopel.irc.privileges`? And why isn't the new location for `Identifier` in `sopel.irc.identifiers`? So many questions!)

Exirel replied:
Long term planning: you can't move things around without causing cyclic import, thus breaking changes. I'm OK with that if you are, but I don't expect you to be (yet).

dgw commented on sopel/tools/target.py:
```diff
@@ -193,13 +213,13 @@ def is_oper(self, nick):
 
         """
-        return self.privileges.get(Identifier(nick), 0) & privileges.OPER
+        nick_id = self.make_identifier(nick)
+        return bool(self.privileges.get(nick_id, 0) & privileges.OPER)
```
OK, this (and the numerous similar changes to other methods below) made me think: Why not just make `Channel.privileges` a `SopelIdentifierMemory`?

Exirel replied:
Hm, I need to check that.

Exirel replied:
Yes, this can be done, at a cost: we need to move `SopelIdentifierMemory` from `sopel.tools` to `sopel.tools.memory`, to prevent circular import.

If you want, I will be happy to do that in that PR, or into a future PR. Whatever you prefer!

Exirel replied:
And this lead me to #2236 and #2237, the latter using `SopelIdentifierMemory` in `Channel`.

dgw commented on sopel/trigger.py:
```diff
@@ -106,30 +110,40 @@ class PreTrigger:
     intent_regex = re.compile('\x01(\\S+) ?(.*)\x01')
 
-    def __init__(self, own_nick, line, url_schemes=None):
+    def __init__(
+        self,
+        own_nick: tools.Identifier,
+        line: str,
+        url_schemes: Optional[Sequence] = None,
+        identifier_factory: IdentifierFactory = tools.Identifier,
+    ):
```
`tools.Identifier` still? Is this a proof that the move hasn't affected backward compatibility? 😂

Exirel replied:
That's the goal here, yes!

Exirel replied:
But truly I should change to `identifiers.Identifier`, would be better.

dgw commented on sopel/trigger.py:
```diff
@@ -106,30 +110,40 @@ class PreTrigger:
             tzinfo=datetime.timezone.utc
         )
         if 'time' in self.tags:
+            # ensure "time" is a string (typecheck)
+            tag_time = self.tags['time'] or ''
```
This will catch if `self.tags['time']` is `False` or `0`, maybe, but not if it's some other kind of object that isn't falsy. What am I missing about the utility of this check?

(Plus these changes feel a bit out of place in this patch.)

Exirel replied:
The answer is in the comment above: it ensures that "time" is a string, for type-checking purpose, because `self.tags` has the following type:

    Dict[str, Optional[str]]

It means that any value in `self.tags` can be either a `str` (what we want), or `None` (what we don't want): no integer, no boolean, nothing else but string and `None`. As a result, all I'm making sure here is that I do have a string in the end.

See? Type. It all makes sense.

dgw reviewed (commented):
I made it through. 😩 Expected to break and finish tomorrow, but nope.

While I'm (understandably?) annoyed by all the extra constructor parameters that had to happen to make this work, I get that it's impossible—or at least, _very_ code-smell-y—to do it another way.

Exirel commented on docs/source/plugin/bot.rst:
```diff
@@ -261,8 +262,8 @@ which provides the following information:
 
-       See :meth:`Identifier.is_nick() <sopel.tools.Identifier.is_nick>` for
-       more information.
+       See :meth:`Identifier.is_nick() <sopel.tools.identifiers.Identifier.is_nick>`
+       for more information.
```
Yes! That's for another PR, that will follow the exact same idea: link Identifiers to the Bot's ISUPPORT parameters.

Exirel commented on sopel/tools/time.py:
```diff
@@ -140,9 +140,10 @@ def _check(zone):
         tz = _check(zone)
-        if not tz:
+        if not tz and zone:
             tz = _check(
-                db.get_nick_or_channel_value(zone, 'timezone'))
+                db.get_nick_value(zone, 'timezone')
+                or db.get_channel_value(zone, 'timezone'))
```
Hm. It was back when I didn't know if I could pass an Identifier maker to SopelDB or not, and I thought: hey, `get_nick_or_channel_value` should only accept an Identifier, but it's functionally identical to call one then the other in that case.

However, this is not true anymore, I can revert that change.

Exirel commented on test/test_coretasks.py:
```diff
@@ -315,6 +315,24 @@ def test_handle_isupport(mockbot):
+    assert mockbot.nick.lower() == 'test{a}'
+    assert str(mockbot.nick) == 'Test[a]'
+
+    mockbot.on_message(
```
Fair enough!

Exirel commented on sopel/bot.py:
```diff
@@ -81,23 +81,28 @@ def __init__(self, config, daemon=False):
         """A mode parser used to parse ``MODE`` messages and modestrings."""
 
-        self.channels = tools.SopelIdentifierMemory()
+        self.channels = tools.SopelIdentifierMemory(
+            identifier_factory=self.make_identifier,
+        )
```
Eeyup. I didn't want to make a breaking change here. However, should you choose to accept it, I can make it mandatory.

And yes, the *TBD* description will contain such information, and the doc will come too!

Exirel commented on sopel/tools/identifiers.py:
```diff
@@ -0,0 +1,171 @@
+        self._lowered = self.casemapping(identifier)
+
+    def lower(self):
+        """Get the RFC 2812-compliant lowercase version of this identifier.
+
+        :return: RFC 2812-compliant lowercase version of the
+                 :py:class:`Identifier` instance
+        :rtype: str
```
Oh yeah, that's part of the "doc overhaul".

Exirel commented on sopel/tools/identifiers.py:
```diff
@@ -0,0 +1,171 @@
+        """
+        return self.casemapping(self)
+
+    @staticmethod
+    def _lower(identifier: str):
+        """Convert an identifier to lowercase per RFC 2812.
+
+        :param str identifier: the identifier (nickname or channel) to convert
+        :return: RFC 2812-compliant lowercase version of ``identifier``
+        :rtype: str
+        """
+        if isinstance(identifier, Identifier):
+            return identifier.lower()
+        return rfc1459_lower(identifier)
+
+    @staticmethod
+    def _lower_swapped(identifier: str):
```
Oh, yeah, right. I know why I didn't do it yet: because I want to move the `deprecated` function somewhere else, if you agree: I think `sopel.deprecated` or `sopel.utils` would be good place for that one. I actually think that `from sopel.deprecated import deprecated` works really nicely!

Exirel commented on sopel/irc/__init__.py:
```diff
@@ -123,6 +124,19 @@ def hostmask(self) -> Optional[str]:
 
+    def make_identifier(self, name: str) -> identifiers.Identifier:
+        """Instantiate an Identifier using the bot's context."""
+        casemapping = identifiers.rfc1459_lower
+
+        if 'CASEMAPPING' in self.isupport:
+            casemapping = {
+                'ascii': identifiers.ascii_lower,
+                'rfc1459': identifiers.rfc1459_lower,
+                'rfc1459-strict': identifiers.rfc1459_strict_lower,
+            }.get(self.isupport['CASEMAPPING'], casemapping)
```
Oh but then I don't even need the `if` in`, I can just use `self.isupport.get('CASEMAPPING')`.

The reason for that `if in` is to have the proper structure in the future for `UTF8MAPPING`:

```
casemapping = identifiers.rfc1459_lower

if 'CASEMAPPING' in self.isupport:
    ...
elif 'UTF8MAPPING' in self.isupport:
    # change default
    casemapping = identifiers.ascii_lower

if 'UTF8MAPPING' in self.isupport:
    # apply the usual one then Precis, in that order
```

But that's for another PR, we can simplify here for now.

Exirel commented on docs/source/api.rst:
```diff
@@ -10,8 +10,16 @@ sopel.tools
+
+.. automodule:: sopel.tools.identifiers
+   :members:
    :private-members: Identifier._lower, Identifier._lower_swapped
```
Oh. Bad rebase conflict.

Exirel commented:
I did a rebase and squashed fixup commits. There was some conflicts a bit complicated to sort out properly, I'm glad I made atomic commit so it was easier to test every single one with py.test and flake8 (except for one commit but heh it's squashed now).

I added one commit to update `SopelIdentifierMemory`'s docstring, as requested (I almost forgot about it, so nice catch!).

What's left is:

* to decide if you want me to move memory classes into their own submodule of `sopel.tools` in that PR or not
* to write the PR's description, that must contain a list of all the breaking changes and things to consider for the upgrade

I won't write a migration script in that PR, that's too big of a job, and I'm not even sure how to properly do it at the moment. I know it's doable, because I had to write a database migration system into one of my application a few years ago that was using SQLAlchemy and Alembic, so it's clearly doable.

Exirel commented:
> What's left is:
> 
> * to decide if you want me to move memory classes into their own submodule of `sopel.tools` in that PR or not
> * to write the PR's description, that must contain a list of all the breaking changes and things to consider for the upgrade

* See PR #2237 
* Done

This is ready for a last round of review.

dgw commented on sopel/irc/__init__.py:
```diff
@@ -47,11 +48,11 @@ class AbstractBot(abc.ABC):
     """Abstract definition of Sopel's interface."""
     def __init__(self, settings):
         # private properties: access as read-only properties
-        self._nick = tools.Identifier(settings.core.nick)
         self._user = settings.core.user
         self._name = settings.core.name
         self._isupport = ISupport()
         self._myinfo = None
+        self._nick = self.make_identifier(settings.core.nick)
```
I don't like moving this to the bottom. The bot's nick is arguably its most important attribute, and it should stay grouped with the other identity-related stuff (user and name).

Exirel replied:
It needs at least `self._isupport` first, so I figured I could put it there, given that it might, in the end, require more than that.

dgw replied:
Maybe "needs" is a bit strong at this point in the process? Sure, `self.isupport` will _exist_ but it won't be populated yet. Is there a meaningful, functional difference between this and the old way (with updated import)?

Exirel replied:
In theory yes: if we change the default value for ISupport's CASEMAPPING, we may end up with a different result, which wasn't the case before. In fact, I should even move that *after* `settings` is set, so `make_identifiers()`  could take advantage of the configuration file.

However, I'm not sure to see the problem here: protected and undocumented attributes, all defined in the same bloc of code, very early in `__init__()`. It's not like there are 100 lines before the `self._nick` here.

dgw commented on sopel/irc/__init__.py:
```diff
@@ -330,12 +343,21 @@ def on_error(self):
+
+        This method exists to update the casemapping rules for the
+        :class:`~sopel.tools.identifiers.Identifier` that represents the bot's
+        nick.
```
```suggestion
        nick, e.g. after ISUPPORT info is received.
```

Good idea to mention _why_ one might want to use this? 😸 Can link to some relevant doc section about ISUPPORT.

I also wonder if, perhaps, the method name should be `_rebuild_nick` since it's not really intended for public use?

Exirel replied:
I was thinking about having that private. However, I also realized that a plugin may want to override coretasks's behavior, or have some way to manage the bot's nick and its rules. This method is idempotent, there are very few risk of anything bad happening (if any), and it doesn't accept any parameter, reducing further the risk of anything wrong done to the bot.

dgw replied:
True enough, I suppose. My only real argument against making it public is that it slightly clutters the docs for new users trying to grok the basics—but then, that's what the other current project (expanding the prose-style docs about writing plugins) is meant to address.

Exirel replied:
> that's what the other current project (expanding the prose-style docs about writing plugins) is meant to address.

Exactly yes!

dgw commented on sopel/modules/adminchannel.py:
```diff
@@ -112,7 +112,7 @@ def kick(bot, trigger):
         reasonidx = 3
     reason = ' '.join(text[reasonidx:])
-    if nick != bot.config.core.nick:
+    if nick != bot.make_identifier(bot.config.core.nick):
```
Some future patch making `bot.config.core.nick` already return an `Identifier` would probably be very helpful. Should you agree, I look forward to (what I expect will be) a discussion about how annoying that would be to implement. 😛

Exirel replied:
This is simply not possible: the configuration is evaluated at startup, before the bot even exists in the first place. This is the string put in the configuration file, and I'd prefer the configuration to stay that way if possible.

dgw replied:
Oof, right. Changing the type after bootstrapping (even though `Identifier` is basically `str`) wouldn't be great.

But I still have another question after looking at the context: Why not just use `if nick != bot.nick`?

Exirel replied:
That's actually a good question, why was it like that? I don't know. I don't think I should change that in *that* PR, but it is worth thinking about it.

dgw commented on test/test_db.py:
```diff
@@ -38,41 +38,33 @@ def teardown_function(function):
 
     # Even if the case is different
-    for test in tests:
-        nick_id = test[0]
-        # still no `create`, because the nick ID should already exist now
-        new_id = db.get_nick_id(Identifier(test[2].upper()))
-        assert nick_id == new_id
+    assert nick_id == db.get_nick_id(variant)
```
Not a huge fan of losing the comments about using/not using the `create` argument. Granted I'm biased because I wrote them, but still… Alternative would be to comment above (where `create=True`) that it defaults to `False`.

dgw commented on test/test_db.py:
```diff
@@ -38,41 +38,33 @@ def teardown_function(function):
 
 
-def test_get_nick_id(db):
+@pytest.mark.parametrize('name, slug, variant', (
+    # Check case insensitive with ASCII only
+    ('Embolalia', 'embolalia', 'eMBOLALIA'),
+    # Ensures case conversion is handled properly
+    ('[][]', '{}{}', '[}{]'),
+    # Unicode, just in case
+    ('EmbölaliÅ', 'embölaliÅ', 'EMBöLALIÅ'),
+))
```
Doesn't switching to this remove the verification that different nicks (that aren't equivalent once casemapped) do indeed get different IDs? I think that's why the test used a loop structure before.

Exirel replied:
The previous version thought that they were the same, because it used `str.lower()`, which is not ascii, rfc1459, nor rfc1459-strict. 😁 

This is the part that is "breaking change", which should affect server that allows UTF-8 nicknames and channel names, which will be solved by a future PR trying to implement something like UTF8MAPPING and what is suggested in RFC 7564 (and its upgraded version RFC 8264).

dgw replied:
Seems I didn't quite get my question across, probably due to attaching this comment to lines in the wrong place.

I meant that there doesn't seem to be any equivalent in the new test code below that matches this old check:

```py
    # Check that each nick ended up with a different id
    assert len(set(test[0] for test in tests)) == len(tests)
```

Each parametrized version of the test is going to start with a fresh `db` object, right? So unless I've missed something (always possible 😅), the test has been downgraded to only making sure that `get_nick_id(somenick, create=True)` retrieves an ID, rather than making sure it creates a _new_ ID when expected.

Put another way, nothing about this new test that I can see makes sure that e.g. the ID for the nick `reblochon`, when seen for the first time, is different from the ID for `brie` (also seen for the first time).

Exirel replied:
Bonus point: I've added some comment about why `create=True`. However, I didn't comment on `create=False`, because I'm not documenting the past here, only what *is* as the moment.

dgw commented on sopel/trigger.py:
```diff
@@ -163,21 +178,25 @@ def __init__(self, own_nick, line, url_schemes=None):
         if self.args and self.event != 'QUIT':
-            target = tools.Identifier(self.args[0])
-        else:
-            target = None
+            target = self.make_identifier(self.args[0])
+
+            # Unless we're messaging the bot directly, in which case that
+            # second arg will be our bot's name.
+            if target.lower() == own_nick.lower():
+                target = self.nick
 
-        # Unless we're messaging the bot directly, in which case that second
-        # arg will be our bot's name.
-        if target and target.lower() == own_nick.lower():
-            target = self.nick
         self.sender = target
```
I've been staring at this for 5 minutes (not counting however long I stared at it in the last review) and… Well, it'll be faster just to ask what you're trying to accomplish with the restructuring. 😅

Exirel replied:
The core idea is that it doesn't make sense to have a second `if target` if we can put that directly within the first `if` that ensure it's not `None`:

1. The target is None by default, with a type hint for an optional Identifier.
2. If it's not a `QUIT` event, it becomes an Identifier (before: it would either become an Identifier or be `None`)
3. If we just created the `target`, we can check if it's the bot's nick or not

The diff is not the most readable things to be honest.

dgw commented on sopel/tools/target.py:
```diff
@@ -61,33 +71,46 @@ class Channel:
-    :type name: :class:`~.tools.Identifier`
+    :type name: :class:`~sopel.tools.identifiers.Identifier`
+    :param identifier_factory: A factory to create
+                               :class:`~sopel.tools.identifiers.Identifier`
```
```suggestion
                               :class:`~sopel.tools.identifiers.Identifier`\\s
```

❔

Also I don't think this is the only place where such a docstring appears.

Exirel replied:
I knew I should have wrote "instances of" and not bother with that `\\s`.

dgw commented on sopel/tools/target.py:
```diff
@@ -265,13 +292,13 @@ def is_admin(self, nick):
 
         """
-        return self.privileges.get(Identifier(nick), 0) & privileges.ADMIN
+        nick_id = self.make_identifier(nick)
+        return bool(self.privileges.get(nick_id, 0) & privileges.ADMIN)
```
Double take alert… I thought this was touching the database (!) for some unknown reason. This might reduce the WTF factor for future developers trekking through this part of the code:

```suggestion
        identifier = self.make_identifier(nick)
        return bool(self.privileges.get(identifier, 0) & privileges.ADMIN)
```

Or one could K.I.S.S. instead (if type-checking doesn't get mad about this):

```suggestion
        nick = self.make_identifier(nick)
        return bool(self.privileges.get(nick, 0) & privileges.ADMIN)
```

_Consider this same note applied to all the methods below using the same local variable name, `nick_id`._

Exirel replied:
Type checking will get extra mad at this, yes.

dgw replied:
So I hear that the type-checker doesn't like changing the type of a variable, fine. First version, then? I'm just looking for an alternative name that isn't `nick_id`. 😺

Exirel replied:
Yup, `identifier` is good!

dgw commented on sopel/tools/__init__.py:
```diff
@@ -620,24 +501,51 @@ class SopelIdentifierMemory(SopelMemory):
 
+    As it works with :class:`~.identifiers.Identifier`, it accepts an
+    identifier factory. This factory usually comes from a
+    :class:`bot instance<sopel.bot.Sopel>`, like in the example of a plugin
```
Low priority: ``:class:`bot instance<sopel.bot.Sopel>`,`` will render "`bot instance`" in fixed-width font (see [deploy preview](https://deploy-preview-2231--sopel-unstable-docs.netlify.app/api.html#sopel.tools.SopelIdentifierMemory)). This is not _broken_ per se, but it's one recurring annoyance that I have with Sphinx docs. Sometimes you want to link a class or function with custom, normal text. 😤

Something like this could work?

```suggestion
    :ref:`bot instance<py:class:sopel.bot.Sopel>`, like in the example of a plugin
```

Exirel replied:
Oh... you are right. I'll try that, otherwise I think I know a way.

dgw commented on sopel/tools/identifiers.py:
```diff
@@ -0,0 +1,266 @@
+
+        .. versionchanged:: 8.0
+
+            Now use the :attr:`casemapping` function to lower the identifier.
```
```suggestion
            Now uses the :attr:`casemapping` function to lower the identifier.

```

dgw commented on sopel/tools/identifiers.py:
```diff
@@ -0,0 +1,266 @@
+
+            This method was added to ensure migration of improperly lowercased
+            data: it reverts the data back to the previous lowercase rules.
+        """
```
```suggestion

        """
```

dgw commented on sopel/tools/identifiers.py:
```diff
@@ -0,0 +1,266 @@
+
+    As per :rfc:`1459#section-2.2`:
+
+        Because of IRC's scandanavian origin, the characters ``{}|`` are
```
I'll be over here, twitching, because I have to leave this error alone in the interest of faithfully quoting the source. 🙈

Exirel replied:
I find it hilarious to be honest.

dgw commented on sopel/tools/identifiers.py:
```diff
@@ -0,0 +1,266 @@
+def rfc1459_strict_lower(text: str) -> str:
+    """Lower ``text`` according to :rfc:`1459` (strict version).
+
+    As per :rfc:`1459#section-2.2`:
```
Anywhere a specific section is linked like this, it would be useful to override the display text, e.g. "RFC 1459 § 2.2", rather than having the `#section-x.y` fragment identifier included in the final docs.

I am, in fact, about to research how annoying it would be to define our own custom role that takes an RFC number _and_ section number, and formats it appropriately without having to muck with manual overrides every time.

Exirel replied:
Oh oh oh, good luck with that. I know what it would take, and I'm not super into it. 😁

dgw replied:
Just fix the rendering for this PR by overriding the display text, and let me worry about scratching the Sphinx-extension itch if I start feeling it really strongly. 😉

In this case, the below would implement the example I wrote above:

```suggestion
    As per :rfc:`RFC 1459 § 2.2<1459#section-2.2>`:
```

Which renders a link to `https://tools.ietf.org/html/rfc1459.html#section-2.2` like this (on Alabaster):

![image](https://user-images.githubusercontent.com/164140/149685303-75623e2e-0251-46dd-8c4e-39e64a4239d1.png)

(Link rendering shouldn't differ between Alabaster and Furo, so I didn't bother changing branches.)

dgw reviewed (commented):
Hm, I found a surprising number of things left to say, despite the many previous comments and despite only re-reviewing about half the files (mostly the ones that "Changed since your last review"). Nobody would ever accuse me of being a code-review Yes Man, huh?

Exirel commented:
Alright this is good, the last review is all about a better presentation and attention to details, I really like it. 👍

Exirel commented:
That should be good now, @dgw you can read my replies and see if I need to take action or if I can rebase & squash away.

dgw commented on test/test_db.py:
```diff
@@ -39,40 +46,66 @@ def teardown_function(function):
+    nick = Identifier('Exirel')
+    session = db.session()
+
+    # Attempt to get NickIDs: it is not created by default
+    with pytest.raises(ValueError):
+        db.get_nick_id(nick)
+
+    # Create the NickIDs
+    nick_id = db.get_nick_id(nick, create=True)
+
+    # Check that one and only one Nicknames exist with that ID
```
I recall now the danger of asking you to write comments sometimes: the nemesi**S** can strike 😹

```suggestion
    # Attempt to get NickID: it is not created by default
    with pytest.raises(ValueError):
        db.get_nick_id(nick)

    # Create the NickID
    nick_id = db.get_nick_id(nick, create=True)

    # Check that one and only one Nickname exists with that ID
```

Exirel replied:
The classes are named `NickIDs` and `Nicknames` tho. 😁

dgw replied:
The test isn't checking for instances of the classes, though. The classes represent database tables, but the test is checking for values. Maybe clearer in the `NickIDs` case to just replace it with `nick ID`, and lowercase `nickname`?

Exirel replied:
Yes, as per our IRC discussion!

dgw commented on test/test_db.py:
```diff
@@ -39,40 +46,66 @@ def teardown_function(function):
+    session = db.session()
+    nick = Identifier(name)
+
+    # Create the NickIDs
```
```suggestion
    # Create the NickID
```

dgw commented on test/test_db.py:
```diff
@@ -39,40 +46,66 @@ def teardown_function(function):
-        assert nick_id == new_id
+    assert nick_id == db.get_nick_id(variant)
+
+    # And no other NickIDs was created (even with create=True)
```
```suggestion
    # And no other NickIDs are created (even with create=True)
```

dgw commented on test/test_db.py:
```diff
@@ -39,40 +46,66 @@ def teardown_function(function):
+    assert nick_id == db.get_nick_id(variant, create=True)
+    assert 1 == session.query(NickIDs).count()
+
+    # But a very different name means a new NickIDs
```
```suggestion
    # But a truly different name means a new NickID
```

Now I've added "splitting semantic hairs" to the ongoing battle against the nemesi**S** 😀

dgw commented on sopel/tools/identifiers.py:
```diff
@@ -0,0 +1,268 @@
+which RFC and how that RFC is implemented by the server: IRC being an old
+protocol, different RFCs have differents version of that process:
+
+* :rfc:`1549#section-2.2`: ASCII characters, and ``[]\\`` are mapped to ``{}|``
+* :rfc:`2812#section-2.2`: same as in the previous RFC, adding ``~`` mapped to
+  ``^``
```
```suggestion
* :rfc:`RFC 1459 § 2.2<1459#section-2.2>`: ASCII characters, and ``[]\\`` are
  mapped to ``{}|``
* :rfc:`RFC 2812 § 2.2<2812#section-2.2>`: same as in the previous RFC, adding
  ``~`` mapped to ``^``
```

Bonus catch: transposed digits. We're not talking about [PPP in HDLC Framing](https://datatracker.ietf.org/doc/html/rfc1549.html). 😄

dgw commented on sopel/tools/identifiers.py:
```diff
@@ -0,0 +1,268 @@
+    """Lower ``text`` according to :rfc:`2812`.
+
+    Similar to :func:`rfc1459_strict_lower`, but also maps ``~`` to ``^``, as
+    per :rfc:`2812#section-2.2`:
```
```suggestion
    per :rfc:`RFC 2812 § 2.2<2812#section-2.2>`:
```

dgw commented on sopel/tools/identifiers.py:
```diff
@@ -0,0 +1,268 @@
+
+    This case insensitivity uses the provided ``casemapping`` function,
+    following the rules for the `CASEMAPPING parameter`__ from ISUPPORT. By
+    default, it uses :func:`rfc1459_lower`, following :rfc:`2812#section-2.2`.
```
```suggestion
    default, it uses :func:`rfc1459_lower`, following
    :rfc:`RFC 2812 § 2.2<2812#section-2.2>`.
```

dgw reviewed (commented):
Just one conversation from before left unresolved—but I started some new ones. Hopefully they won't make you 🙁 too much, since it's about fixing details & consistency in the new changes.

Exirel commented:
And once again! But this time, no new docstrings. :grin:

dgw commented:
I looked, all good. Squash what you're going to squash when ready.

dgw reviewed (approved):
I'm going to do what I should have done earlier and actually approve this.

You can squash the fixup into the commit SHA I gave and trust me (and CI) that everything will work out post-merge, or rebase on `master` if you like.
